lib/puppet/http/resolver/serverlist.rb

```diff
@@ -0,0 +1,30 @@
+class Puppet::HTTP::Resolver::ServerList < Puppet::HTTP::Resolver
+  def initialize(client, server_list: Puppet[:server_list], default_port: Puppet[:masterport])
+    @client = client
+    @server_list = server_list
+    @default_port = default_port
+  end
+
+  def resolve(session, name, &block)
+    success = false
+    @server_list.each do |server|
+      host = server[0]
+      port = server[1] || @default_port
+      begin
+        response = @client.get(URI("https://#{host}:#{port}/status/v1/simple/master"))
+        if response.success?
+          success = true
+          yield session.create_service(name, host, port)
+        end
+
+        Puppet.debug(_("Puppet server %{host}:%{port} is unavailable: %{code} %{reason}") %
+                     { host: host, port: port, code: response.code, reason: response.message })
+      rescue => detail
+        #TRANSLATORS 'server_list' is the name of a setting and should not be translated
+        Puppet.debug _("Unable to connect to server from server_list setting: %{detail}") % {detail: detail}
+      end
+    end
+
+    raise Puppet::Error, _("Could not select a functional puppet master from server_list: '%{server_list}'") % { server_list: Puppet.settings.value(:server_list, Puppet[:environment].to_sym, true) } unless success
```

Comment:
I'm thinking this is a policy decision, which should reside in the session. If the session calls the server list resolver, but there are no functional services, then the session should raise.

Reply:
we decided it was okay to keep this policy decision here for the moment, with the idea that it'll move elsewhere in the future